mempool/clist_sidecar.go

```diff
@@ -0,0 +1,655 @@
+		if _, loaded := orderedTxsMap.LoadOrStore(txInfo.BundleOrder, scTx); loaded {
+			// if we had the tx already, then skip
+			sc.logger.Info(
+				"failed adding sidecarTx",
+				"reason", "already have a tx for this BundleID, height, and bundleOrder...",
+				"bundle id", txInfo.BundleID,
+				"bundle height", txInfo.DesiredHeight,
+				"bundle order", txInfo.BundleOrder,
+				"tx", tx.Hash(),
+			)
+			return true, nil
+		}
+		// if we added, then increment bundle size for BundleID
+		atomic.AddInt64(&bundle.CurrSize, int64(1))
+		return false, nil
+	}()
+	if err != nil {
+		return err
+	} else if shouldReturn {
+		return nil
+	}
+
+	// -------- UPDATE MAX BUNDLE ---------
+
+	func() {
+		sc.maxBundleIDMtx.Lock()
+		defer sc.maxBundleIDMtx.Unlock()
+		if txInfo.BundleID >= sc.maxBundleID {
+			sc.maxBundleID = txInfo.BundleID
+		}
+	}()
+
+	// -------- TX INSERTION INTO MAIN TXS LIST ---------
+	// -------- TODO: In the future probably want to refactor to not have txs clist ---------
+
+	e := sc.txs.PushBack(scTx)
+	sc.txsMap.Store(scTx.Tx.Key(), e)
+	atomic.AddInt64(&sc.txsBytes, int64(len(scTx.Tx)))
+
+	// add metric for new sidecar size
+	sc.metrics.MevBundleMempoolSize.Set(float64(sc.Size()))
+
+	// add metric for sidecar tx sampling
+	sc.metrics.MevTxSizeBytes.Observe(float64(len(scTx.Tx)))
+
+	sc.lastBundleHeight = scTx.DesiredHeight
+
+	// TODO: in the future, refactor to only notifyTxsAvailable when we have at least one full bundle
+	if sc.Size() > 0 {
+		sc.notifyTxsAvailable()
+	}
+	sc.logger.Info(
+		"Added sidecar tx successfully",
+		"tx", tx,
+		"BundleID", txInfo.BundleID,
+		"bundleOrder", txInfo.BundleOrder,
+		"desiredHeight", txInfo.DesiredHeight,
+		"bundleSize", txInfo.BundleSize,
+		"sidecar total size", sc.Size(),
+	)
+
+	return nil
+}
+
+// Gets the height of the last received bundle tx. For status rpc purposes.
+func (sc *CListPriorityTxSidecar) GetLastBundleHeight() int64 {
+	if sc == nil {
+		return 0
+	}
+	return sc.lastBundleHeight
+}
+
+// TxsWaitChan returns a channel to wait on transactions. It will be closed
+// once the sidecar is not empty (ie. the internal `mem.txs` has at least one
+// element)
+//
+// Safe for concurrent use by multiple goroutines.
+func (sc *CListPriorityTxSidecar) TxsWaitChan() <-chan struct{} {
+	return sc.txs.WaitChan()
+}
+
+// TxsFront returns the first transaction in the ordered list for peer
+// goroutines to call .NextWait() on.
+// FIXME: leaking implementation details!
+//
+// Safe for concurrent use by multiple goroutines.
+func (sc *CListPriorityTxSidecar) TxsFront() *clist.CElement {
+	return sc.txs.Front()
+}
+
+// Lock() must be held by the caller during execution.
+func (sc *CListPriorityTxSidecar) Update(
+	height int64,
+	txs types.Txs,
+	deliverTxResponses []*abci.ResponseDeliverTx,
+) error {
+
+	// Set height for block last updated to (i.e. block last committed)
+	sc.height = height
```

Comment:
Not used elsewhere

Reply:
Listed in https://github.com/informalsystems/audit-skip/issues/6

Reply:
Removed in https://github.com/skip-mev/mev-tendermint/pull/126